presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -739,18 +742,78 @@ protected Optional<? extends TableEncryptionProperties> getTableEncryptionProper
-            return Optional.of(new HiveInputInfo(
-                    tableLayoutHandle.getPartitions().get().stream()
-                            .map(HivePartition::getPartitionId)
-                            .collect(toList()),
-                    false));
+        if (!tableLayoutHandle.getPartitions().isPresent()
+                || 0 == tableLayoutHandle.getPartitions().get().size()) {
+            return Optional.empty();
         }
 
-        return Optional.empty();
+        MetastoreContext metastoreContext = new MetastoreContext(
+                session.getIdentity(),
+                session.getQueryId(),
+                session.getClientInfo(),
+                session.getSource(),
+                getMetastoreHeaders(session),
+                isUserDefinedTypeEncodingEnabled(session),
+                metastore.getColumnConverterProvider());
+
+        if (isUnpartitionedTable(tableLayoutHandle)) {
+            return getHiveInputInfoForUnpartitionedTable(metastoreContext, tableLayoutHandle);
+        }
+        return getHiveInputInfoForPartitionedTable(metastoreContext, tableLayoutHandle);
+    }
+
+    private boolean isUnpartitionedTable(HiveTableLayoutHandle tableLayoutHandle)
+    {
+        List<HivePartition> partitions = tableLayoutHandle.getPartitions().get();
+        if (1 == partitions.size()
+                && 0 == partitions.get(0).getPartitionId().compareTo(DUMMY_PARTITION_NAME)) {
+            return true;
+        }
+        return false;
+    }
+
+    private Optional<Object> getHiveInputInfoForUnpartitionedTable(
+            MetastoreContext metastoreContext, HiveTableLayoutHandle tableLayoutHandle)
+    {
+        Optional<Table> table = metastore.getTable(
+                metastoreContext,
+                tableLayoutHandle.getSchemaTableName().getSchemaName(),
+                tableLayoutHandle.getSchemaTableName().getTableName());
+
+        int lastAccessTime = 0;
+        if (table.isPresent()) {
+            lastAccessTime = table.get().getLastAccessTime();
+        }
+        return Optional.of(new HiveInputInfo(
+                ImmutableList.of(DUMMY_PARTITION_NAME),
+                false,
+                ImmutableList.of(lastAccessTime)));
+    }
+
+    private Optional<Object> getHiveInputInfoForPartitionedTable(
+            MetastoreContext metastoreContext, HiveTableLayoutHandle tableLayoutHandle)
+    {
+        List<String> partitionNames = tableLayoutHandle
+                .getPartitions().get().stream()
+                .map(HivePartition::getPartitionId)
+                .limit(MAX_NUM_OF_FETCHED_PARTITIONS)
+                .collect(toList());
+
+        Map<String, Optional<Partition>> partitions = metastore.getPartitionsByNames(
+                metastoreContext,
+                tableLayoutHandle.getSchemaTableName().getSchemaName(),
+                tableLayoutHandle.getSchemaTableName().getTableName(),
+                partitionNames);
```

Comment:
The number of partitions could be very large. Where the data will be stored? It could take lots of storage. Should we add a cap for the number of partitions?

How will this affect the query latency? Maybe we need to do some replay to figure it out.

Reply:
Agree with this.  I think we need to be careful here for a couple reasons
1. getPartitionsByNames() can be expensive on the metastore side if you are querying lots of partitions
2. If there are lots of partitions, keeping the last access times for all of them can put extra memory pressure on the coordinator

It would be nice if we could use the last access times from other existing metastore calls that we already make to get partitions  (e.g. when we get the partitions for the split info), but I haven't looked into how feasible that is.

At the very least I would do some testing to make sure that this doesn't cause issues (for the cluster or the query) if you are running several queries that access a lot of partitions.  I would also make sure that the partition limit is configurable and maybe add a flag to disable getting lastAccessTimes altogether.

Reply:
+1; still recommend to have a meeting to discuss this. The metastore change looks too complicated that introduces a lot of new overheads.

Reply:
Yeah, let us discuss offline for this.